scripts/bert/run_pretraining_hvd.py

```diff
@@ -0,0 +1,235 @@
+                    # if accumulate > 1, grad_req is set to 'add', and zero_grad is required
+                    if accumulate > 1:
+                        param_dict.zero_grad()
+                    # update learning rate
+                    if step_num <= num_warmup_steps:
+                        new_lr = lr * step_num / num_warmup_steps
+                    else:
+                        offset = lr * step_num / num_train_steps
+                        new_lr = lr - offset
+                    trainer.set_learning_rate(new_lr)
+                    if args.profile:
+                        profile(step_num, 10, 14, profile_name=args.profile + str(rank))
+
+                # load data
+                if args.use_avg_len:
+                    data_list = [[seq.as_in_context(context) for seq in shard]
+                                 for context, shard in zip([ctx], data_batch)]
+                else:
+                    data_list = split_and_load(data_batch, [ctx])
+                data = data_list[0]
+
+                # forward
+                with mx.autograd.record():
+                    (ls, ns_label, classified, masked_id, decoded, \
+                     masked_weight, ls1, ls2, valid_len) = forward(data, model, mlm_loss,
+                                                                   nsp_loss, vocab_size, args.dtype)
+                    ls = ls / accumulate
+                    # backward
+                    if args.dtype == 'float16':
+                        fp16_trainer.backward(ls)
+                    else:
+                        ls.backward()
+
+                running_mlm_loss += ls1.as_in_context(mx.cpu())
+                running_nsp_loss += ls2.as_in_context(mx.cpu())
+                running_num_tks += valid_len.sum().as_in_context(mx.cpu())
+
+                # update
+                if (batch_num + 1) % accumulate == 0:
+                    fp16_trainer.step(1, max_norm=1)
+
+                nsp_metric.update([ns_label], [classified])
+                mlm_metric.update([masked_id], [decoded], [masked_weight])
+
+                # logging
+                if (step_num + 1) % (args.log_interval) == 0 and (batch_num + 1) % accumulate == 0:
+                    log(begin_time, running_num_tks, running_mlm_loss / accumulate,
+                        running_nsp_loss / accumulate, step_num, mlm_metric, nsp_metric,
+                        trainer, args.log_interval)
+                    begin_time = time.time()
+                    running_mlm_loss = running_nsp_loss = running_num_tks = 0
+                    mlm_metric.reset_local()
+                    nsp_metric.reset_local()
+
+                # saving checkpoints
+                if args.ckpt_dir and (step_num + 1) % (args.ckpt_interval) == 0 \
+                   and (batch_num + 1) % accumulate == 0 and local_rank == 0:
+                    save_params(step_num, model, trainer, args.ckpt_dir)
+
+                batch_num += 1
+
+    if local_rank == 0:
```

Comment:
Save checkpoints only on worker 0 to prevent other workers from corrupting them. if run a job on 2 machines, and each machine has 4 GPUs, the rank will be a number [0, 7], and the local rank will be a number [0, 3]. So it should be saved when rank is 0.
```python
if args.ckpt_dir and rank == 0
```

Reply:
I wanted to make sure all machines have the checkpoint, as the checkpoint has to be present at each machine when loading it. There will be 2 processes with local_rank = 0, and only 1 process with rank = 0.

Reply:
Ok, got it.